File: android/src/main/kotlin/ch/waio/pro_video_editor/src/features/render/helpers/VideoTranscoder.kt
package ch.waio.pro_video_editor.src.features.render.helpers

import RENDER_TAG
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.media3.common.MediaItem
import androidx.media3.common.MimeTypes
import androidx.media3.common.util.UnstableApi
import androidx.media3.transformer.Composition
import androidx.media3.transformer.DefaultEncoderFactory
import androidx.media3.transformer.EditedMediaItem
import androidx.media3.transformer.EditedMediaItemSequence
import androidx.media3.transformer.ExportException
import androidx.media3.transformer.ExportResult
import androidx.media3.transformer.Transformer
import java.io.File
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicReference

/**
 * Utility for pre-transcoding HEVC 10-bit HDR videos to H.264 8-bit.
 * 
 * This is necessary because GPU-based effect pipelines on Android have compatibility
 * issues with HEVC Main 10 Profile (hvc1.2.4.H120) videos. By transcoding to H.264 first,
 * we can then safely apply effects like ColorMatrix, Blur, or Overlay.
 */
@UnstableApi
object VideoTranscoder {

    /**
     * Result of a transcoding operation.
     */
    sealed class TranscodeResult {
        /** Transcoding succeeded, contains path to transcoded file */
        data class Success(val outputPath: String) : TranscodeResult()

        /** No transcoding needed, original file is compatible */
        data class NotNeeded(val originalPath: String) : TranscodeResult()

        /** Transcoding failed with error */
        data class Error(val exception: Throwable) : TranscodeResult()
    }

    /**
     * Checks if a video needs transcoding for effect compatibility.
     * 
     * @param videoPath Path to the video file
     * @return True if transcoding is needed
     */
    fun needsTranscoding(videoPath: String): Boolean {
        val formatInfo = MediaInfoExtractor.getVideoFormatInfo(videoPath)
        val needsTranscode = formatInfo.needsTranscodingForEffects()

        Log.d(
            RENDER_TAG, "Video transcoding check: path=$videoPath, " +
                    "isHevc=${formatInfo.isHevc}, bitDepth=${formatInfo.bitDepth}, " +
                    "isHdr=${formatInfo.isHdr}, needsTranscoding=$needsTranscode"
        )

        return needsTranscode
    }

    /**
     * Transcodes a video to H.264 8-bit SDR format for effect compatibility.
     * 
     * Uses HDR -> SDR tonemapping to convert 10-bit HDR to 8-bit SDR,
     * which then allows H.264 encoding.
     * 
     * This is a blocking operation that should be called from a background thread.
     * The transcoded file is saved to the app's cache directory.
     * 
     * @param context Android context
     * @param inputPath Path to the input video
     * @return TranscodeResult indicating success, not-needed, or error
     */
    fun transcodeToH264Sync(context: Context, inputPath: String): TranscodeResult {
        // Check if transcoding is needed
        if (!needsTranscoding(inputPath)) {
            Log.d(RENDER_TAG, "No transcoding needed for: $inputPath")
            return TranscodeResult.NotNeeded(inputPath)
        }

        Log.i(RENDER_TAG, "Starting HEVC 10-bit HDR -> H.264 8-bit SDR transcoding for: $inputPath")

        val outputFile = File(
            context.cacheDir,
            "transcoded_${System.currentTimeMillis()}.mp4"
        )

        val resultRef = AtomicReference<TranscodeResult>()
        val latch = CountDownLatch(1)
        val mainHandler = Handler(Looper.getMainLooper())

        mainHandler.post {
            try {
                // Create encoder factory that forces H.264
                val encoderFactory = DefaultEncoderFactory.Builder(context)
                    .setEnableFallback(true)
                    .build()

                val transformer = Transformer.Builder(context)
                    .setVideoMimeType(MimeTypes.VIDEO_H264)  // Force H.264 output
                    .setEncoderFactory(encoderFactory)
                    .addListener(object : Transformer.Listener {
                        override fun onCompleted(composition: Composition, result: ExportResult) {
                            Log.i(RENDER_TAG, "Transcoding completed: ${outputFile.absolutePath}")

                            // Verify the output is actually H.264
                            val outputInfo =
                                MediaInfoExtractor.getVideoFormatInfo(outputFile.absolutePath)
                            Log.i(
                                RENDER_TAG, "Transcoded output: isHevc=${outputInfo.isHevc}, " +
                                        "bitDepth=${outputInfo.bitDepth}, isHdr=${outputInfo.isHdr}"
                            )

                            resultRef.set(TranscodeResult.Success(outputFile.absolutePath))
                            latch.countDown()
                        }

                        override fun onError(
                            composition: Composition,
                            result: ExportResult,
                            exception: ExportException
                        ) {
                            Log.e(RENDER_TAG, "Transcoding failed: ${exception.message}")
                            outputFile.delete()
                            resultRef.set(TranscodeResult.Error(exception))
                            latch.countDown()
                        }
                    })
                    .build()

                // Create composition with HDR tonemapping to force SDR output
                val mediaItem = MediaItem.Builder()
                    .setUri(inputPath)
                    .build()

                // Use HDR_MODE_TONE_MAP_HDR_TO_SDR_USING_OPEN_GL to convert HDR to SDR
                // This forces 8-bit output which then allows H.264 encoding
                val editedMediaItem = EditedMediaItem.Builder(mediaItem)
                    .setRemoveAudio(false)
                    .setRemoveVideo(false)
                    .build()

                // Build composition with HDR tonemapping enabled
                val sequence = EditedMediaItemSequence.Builder(editedMediaItem).build()
                val composition = Composition.Builder(sequence)
                    // Force HDR to SDR conversion - this enables H.264 encoding
                    .setHdrMode(Composition.HDR_MODE_TONE_MAP_HDR_TO_SDR_USING_OPEN_GL)
                    .build()

                transformer.start(composition, outputFile.absolutePath)

            } catch (e: Exception) {
                Log.e(RENDER_TAG, "Failed to start transcoding: ${e.message}")
                resultRef.set(TranscodeResult.Error(e))
                latch.countDown()
            }
        }

        // Wait for transcoding to complete
        try {
            latch.await()
        } catch (e: InterruptedException) {
            return TranscodeResult.Error(e)
        }

        return resultRef.get() ?: TranscodeResult.Error(
            IllegalStateException("Transcoding result not set")
        )
    }

    /**
     * Async version of transcoding.
     * 
     * @param context Android context
     * @param inputPath Path to the input video
     * @param onComplete Callback with result
     */
    fun transcodeToH264Async(
        context: Context,
        inputPath: String,
        onComplete: (TranscodeResult) -> Unit
    ) {
        Thread {
            val result = transcodeToH264Sync(context, inputPath)
            Handler(Looper.getMainLooper()).post {
                onComplete(result)
            }
        }.start()
    }

    /**
     * Transcodes multiple video clips if needed.
     * 
     * @param context Android context
     * @param inputPaths List of input video paths
     * @return Map of original path to transcoded path (or original if no transcoding needed)
     */
    fun transcodeClipsIfNeeded(
        context: Context,
        inputPaths: List<String>
    ): Map<String, String> {
        val result = mutableMapOf<String, String>()

        for (inputPath in inputPaths) {
            when (val transcodeResult = transcodeToH264Sync(context, inputPath)) {
                is TranscodeResult.Success -> {
                    result[inputPath] = transcodeResult.outputPath
                }

                is TranscodeResult.NotNeeded -> {
                    result[inputPath] = transcodeResult.originalPath
                }

                is TranscodeResult.Error -> {
                    Log.e(RENDER_TAG, "Transcoding failed for $inputPath, using original")
                    result[inputPath] = inputPath
                }
            }
        }

        return result
    }

    /**
     * Cleans up transcoded temporary files.
     * 
     * @param transcodedPaths List of transcoded file paths to delete
     */
    fun cleanupTranscodedFiles(transcodedPaths: Collection<String>) {
        for (path in transcodedPaths) {
            val file = File(path)
            if (file.exists() && file.absolutePath.contains("transcoded_")) {
                val deleted = file.delete()
                Log.d(RENDER_TAG, "Cleanup transcoded file: $path, deleted=$deleted")
            }
        }
    }
}

File: macos/Classes/src/features/render/helpers/VideoTranscoder.swift
import AVFoundation
import CoreImage
import Foundation

/// Utility for pre-transcoding HEVC 10-bit HDR videos to H.264 8-bit SDR.
///
/// This is necessary because GPU-based effect pipelines on macOS have compatibility
/// issues with HEVC Main 10 Profile videos. By transcoding to H.264 first,
/// we can then safely apply effects like ColorMatrix, Blur, or Overlay.
internal class VideoTranscoder {

    // MARK: - Result Types

    /// Result of a transcoding operation.
    enum TranscodeResult {
        /// Transcoding succeeded, contains path to transcoded file
        case success(outputPath: String)

        /// No transcoding needed, original file is compatible
        case notNeeded(originalPath: String)

        /// Transcoding failed with error
        case error(Error)
    }

    // MARK: - Public Methods

    /// Checks if a video needs transcoding for effect compatibility.
    ///
    /// - Parameter videoPath: Path to the video file
    /// - Returns: True if transcoding is needed
    static func needsTranscoding(_ videoPath: String) async -> Bool {
        let formatInfo = await MediaInfoExtractor.getVideoFormatInfo(videoPath)
        let needsTranscode = formatInfo.needsTranscodingForEffects()

        PluginLog.print(
            "🔍 Video transcoding check: path=\(videoPath), "
                + "isHevc=\(formatInfo.isHevc), bitDepth=\(formatInfo.bitDepth), "
                + "isHdr=\(formatInfo.isHdr), needsTranscoding=\(needsTranscode)")

        return needsTranscode
    }

    /// Transcodes a video to H.264 8-bit SDR format for effect compatibility.
    ///
    /// Uses HDR → SDR tonemapping to convert 10-bit HDR to 8-bit SDR,
    /// which allows proper GPU effect processing.
    ///
    /// - Parameter videoPath: Path to the input video
    /// - Returns: TranscodeResult indicating success, not-needed, or error
    static func transcodeToH264(_ videoPath: String) async -> TranscodeResult {
        // Check if transcoding is needed
        guard await needsTranscoding(videoPath) else {
            PluginLog.print("✅ No transcoding needed for: \(videoPath)")
            return .notNeeded(originalPath: videoPath)
        }

        PluginLog.print("🎬 Starting HEVC 10-bit HDR → H.264 8-bit SDR transcoding for: \(videoPath)")

        let inputURL = URL(fileURLWithPath: videoPath)
        let outputURL = FileManager.default.temporaryDirectory
            .appendingPathComponent("transcoded_\(Int(Date().timeIntervalSince1970 * 1000)).mp4")

        do {
            try await transcodeVideo(from: inputURL, to: outputURL)

            // Verify output
            let outputInfo = await MediaInfoExtractor.getVideoFormatInfo(outputURL.path)
            PluginLog.print("✅ Transcoding completed: \(outputURL.path)")
            PluginLog.print(
                "   Output: isHevc=\(outputInfo.isHevc), bitDepth=\(outputInfo.bitDepth), isHdr=\(outputInfo.isHdr)"
            )

            return .success(outputPath: outputURL.path)

        } catch {
            PluginLog.print("❌ Transcoding failed: \(error.localizedDescription)")
            try? FileManager.default.removeItem(at: outputURL)
            return .error(error)
        }
    }

    /// Transcodes multiple video clips if needed.
    ///
    /// - Parameter inputPaths: List of input video paths
    /// - Returns: Dictionary mapping original path to transcoded path (or original if no transcoding needed)
    static func transcodeClipsIfNeeded(_ inputPaths: [String]) async -> [String: String] {
        var result: [String: String] = [:]

        for inputPath in inputPaths {
            switch await transcodeToH264(inputPath) {
            case .success(let outputPath):
                result[inputPath] = outputPath
            case .notNeeded(let originalPath):
                result[inputPath] = originalPath
            case .error:
                PluginLog.print("⚠️ Transcoding failed for \(inputPath), using original")
                result[inputPath] = inputPath
            }
        }

        return result
    }

    /// Cleans up transcoded temporary files.
    ///
    /// - Parameter transcodedPaths: Collection of transcoded file paths to delete
    static func cleanupTranscodedFiles(_ transcodedPaths: [String]) {
        for path in transcodedPaths {
            if path.contains("transcoded_") {
                do {
                    try FileManager.default.removeItem(atPath: path)
                    PluginLog.print("🗑️ Cleaned up transcoded file: \(path)")
                } catch {
                    PluginLog.print("⚠️ Failed to clean up \(path): \(error.localizedDescription)")
                }
            }
        }
    }

    // MARK: - Private Methods

    /// Performs the actual video transcoding using AVAssetExportSession.
    /// This is simpler and more reliable than AVAssetWriter for basic transcoding.
    private static func transcodeVideo(from inputURL: URL, to outputURL: URL) async throws {
        let asset = AVURLAsset(
            url: inputURL,
            options: [
                AVURLAssetPreferPreciseDurationAndTimingKey: true
            ])

        // Use AVAssetExportSession for simpler, more reliable transcoding
        guard
            let exportSession = AVAssetExportSession(
                asset: asset, presetName: AVAssetExportPresetHighestQuality)
        else {
            throw NSError(
                domain: "VideoTranscoder", code: 1,
                userInfo: [NSLocalizedDescriptionKey: "Failed to create export session"])
        }

        exportSession.shouldOptimizeForNetworkUse = true
        if #unavailable(macOS 15.0) {
            exportSession.outputURL = outputURL
            exportSession.outputFileType = .mp4
        }

        // Create video composition for HDR → SDR conversion
        let videoTrack: AVAssetTrack
        if #available(macOS 13.0, *) {
            let videoTracks = try await asset.loadTracks(withMediaType: .video)
            guard let vTrack = videoTracks.first else {
                throw NSError(
                    domain: "VideoTranscoder", code: 2,
                    userInfo: [NSLocalizedDescriptionKey: "No video track found"])
            }
            videoTrack = vTrack
        } else {
            guard let vTrack = asset.tracks(withMediaType: .video).first else {
                throw NSError(
                    domain: "VideoTranscoder", code: 2,
                    userInfo: [NSLocalizedDescriptionKey: "No video track found"])
            }
            videoTrack = vTrack
        }

        // Get video properties
        let naturalSize: CGSize
        let preferredTransform: CGAffineTransform
        let nominalFrameRate: Float

        if #available(macOS 13.0, *) {
            naturalSize = try await videoTrack.load(.naturalSize)
            preferredTransform = try await videoTrack.load(.preferredTransform)
            nominalFrameRate = try await videoTrack.load(.nominalFrameRate)
        } else {
            naturalSize = videoTrack.naturalSize
            preferredTransform = videoTrack.preferredTransform
            nominalFrameRate = videoTrack.nominalFrameRate
        }

        // Calculate render size accounting for rotation
        let renderSize = calculateOutputSize(
            naturalSize: naturalSize, transform: preferredTransform)

        // Create video composition with SDR color space filter
        if #available(macOS 26.0, *) {
            let videoComposition = try await AVVideoComposition(applyingFiltersTo: asset) { params in
                var image = params.sourceImage.clampedToExtent()

                if let colorMatrix = CIFilter(name: "CIColorMatrix") {
                    colorMatrix.setValue(image, forKey: kCIInputImageKey)
                    colorMatrix.setValue(CIVector(x: 1, y: 0, z: 0, w: 0), forKey: "inputRVector")
                    colorMatrix.setValue(CIVector(x: 0, y: 1, z: 0, w: 0), forKey: "inputGVector")
                    colorMatrix.setValue(CIVector(x: 0, y: 0, z: 1, w: 0), forKey: "inputBVector")
                    colorMatrix.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")

                    if let output = colorMatrix.outputImage {
                        image = output
                    }
                }

                let cropped = image.cropped(to: CGRect(origin: .zero, size: renderSize))
                return AVCIImageFilteringResult(resultImage: cropped, ciContext: nil)
            }
            exportSession.videoComposition = videoComposition
        } else {
            let videoComposition = AVMutableVideoComposition(asset: asset) { request in
                var image = request.sourceImage.clampedToExtent()

                if let colorMatrix = CIFilter(name: "CIColorMatrix") {
                    colorMatrix.setValue(image, forKey: kCIInputImageKey)
                    colorMatrix.setValue(CIVector(x: 1, y: 0, z: 0, w: 0), forKey: "inputRVector")
                    colorMatrix.setValue(CIVector(x: 0, y: 1, z: 0, w: 0), forKey: "inputGVector")
                    colorMatrix.setValue(CIVector(x: 0, y: 0, z: 1, w: 0), forKey: "inputBVector")
                    colorMatrix.setValue(CIVector(x: 0, y: 0, z: 0, w: 1), forKey: "inputAVector")

                    if let output = colorMatrix.outputImage {
                        image = output
                    }
                }

                let cropped = image.cropped(to: CGRect(origin: .zero, size: renderSize))
                request.finish(with: cropped, context: nil)
            }

            videoComposition.renderSize = renderSize
            videoComposition.frameDuration = CMTime(value: 1, timescale: CMTimeScale(nominalFrameRate))

            videoComposition.colorPrimaries = AVVideoColorPrimaries_ITU_R_709_2
            videoComposition.colorTransferFunction = AVVideoTransferFunction_ITU_R_709_2
            videoComposition.colorYCbCrMatrix = AVVideoYCbCrMatrix_ITU_R_709_2

            exportSession.videoComposition = videoComposition
        }

        PluginLog.print("🎬 Transcoding with AVAssetExportSession...")
        PluginLog.print("   Input size: \(naturalSize), Output size: \(renderSize)")

        // Export
        if #available(macOS 15.0, *) {
            try await exportSession.export(to: outputURL, as: .mp4)
        } else {
            await exportSession.export()
            guard exportSession.status == .completed else {
                let errorMessage = exportSession.error?.localizedDescription ?? "Unknown error"
                throw NSError(
                    domain: "VideoTranscoder", code: 3,
                    userInfo: [NSLocalizedDescriptionKey: "Export failed: \(errorMessage)"])
            }
        }

        PluginLog.print("✅ Transcoding completed successfully")
    }

    /// Calculates output size accounting for rotation.
    private static func calculateOutputSize(naturalSize: CGSize, transform: CGAffineTransform)
        -> CGSize
    {
        let rotationAngle = atan2(transform.b, transform.a)
        let radians = abs(rotationAngle)

        // Check if rotation is ~90° or ~270°
        if radians > .pi / 4 && radians < 3 * .pi / 4 {
            return CGSize(width: naturalSize.height, height: naturalSize.width)
        }
        return naturalSize
    }

    /// Calculates appropriate bitrate based on resolution and frame rate.
    private static func calculateBitrate(size: CGSize, frameRate: Float) -> Int {
        let pixels = Int(size.width * size.height)
        let fps = max(24, min(60, Int(frameRate)))

        // Roughly 0.1 bits per pixel per frame for H.264 High profile
        return max(2_000_000, pixels * fps / 10)
    }
}
